In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df = pd.read_csv('transaction_clean.csv')

In [3]:
df

Unnamed: 0  UserId  TransactionId               TransactionTime  \
0                0  278166        6355745  Sat Feb 02 12:50:00 IST 2019   
1                1  278166        6355745  Sat Feb 02 12:50:00 IST 2019   
2                2  278166        6355745  Sat Feb 02 12:50:00 IST 2019   
3                3  278166        6355745  Sat Feb 02 12:50:00 IST 2019   
4                4  278166        6355745  Sat Feb 02 12:50:00 IST 2019   
...            ...     ...            ...                           ...   
525883      525883  302484        5953871  Fri Mar 30 11:48:00 IST 2018   
525884      525884  376929        6259946  Wed Dec 12 07:58:00 IST 2018   
525885      525885  339927        6231192  Sat Nov 24 14:19:00 IST 2018   
525886      525886  297297        5995748  Fri May 11 11:07:00 IST 2018   
525887      525887  273441        5918825  Tue Feb 20 10:12:00 IST 2018   

        ItemCode                   ItemDescription  NumberOfItemsPurchased  \
0         465549  FAMILY ALBUM WHITE PICTURE FRAME                       6   
1         487494    JINGLE BELL HEART ANTIQUE GOLD                       3   
2         490686    SET 12 COLOURING PENCILS DOILY                       3   
3         460593   PINK VINTAGE PAISLEY PICNIC BAG                       3   
4         495201            JUMBO BAG PAISLEY PARK                      30   
...          ...                               ...                     ...   
525883    471996  EMPIRE UNION JACK TV DINNER TRAY                     -24   
525884    466137                    POPCORN HOLDER                    3000   
525885    998886                     PARTY BUNTING                     -24   
525886   1769187    FOLK ART GREETING CARD,pack/12                     -33   
525887    446964           RETROSPOT CANDLE  SMALL                     -36   

        CostPerItem         Country  
0             11.73  United Kingdom  
1              2.88  United Kingdom  
2              0.90  United Kingdom  
3              2.28  United Kingdom  
4              2.88  United Kingdom  
...             ...             ...  
525883         6.84  United Kingdom  
525884         1.00  United Kingdom  
525885         6.84  United Kingdom  
525886         4.08  United Kingdom  
525887         2.34  United Kingdom  

[525888 rows x 9 columns]

In [4]:
#there are 8635 negative transactions when we compare it to the 500k rows its around 0.01% of the data o lets drop it
sum(df['NumberOfItemsPurchased'] < 0)

8635

In [5]:
negatives = df[df['NumberOfItemsPurchased'] < 0].index
negatives

Int64Index([  6424,   6425,   6426,   6427,  14769,  14770,  14771,  14772,
             14773,  14774,
            ...
            525874, 525875, 525876, 525877, 525880, 525881, 525883, 525885,
            525886, 525887],
           dtype='int64', length=8635)

In [6]:
df.drop(negatives,inplace=True)

In [7]:
df.drop('Unnamed: 0',axis=1,inplace=True)

In [8]:
df.head()

UserId  TransactionId               TransactionTime  ItemCode  \
0  278166        6355745  Sat Feb 02 12:50:00 IST 2019    465549   
1  278166        6355745  Sat Feb 02 12:50:00 IST 2019    487494   
2  278166        6355745  Sat Feb 02 12:50:00 IST 2019    490686   
3  278166        6355745  Sat Feb 02 12:50:00 IST 2019    460593   
4  278166        6355745  Sat Feb 02 12:50:00 IST 2019    495201   

                    ItemDescription  NumberOfItemsPurchased  CostPerItem  \
0  FAMILY ALBUM WHITE PICTURE FRAME                       6        11.73   
1    JINGLE BELL HEART ANTIQUE GOLD                       3         2.88   
2    SET 12 COLOURING PENCILS DOILY                       3         0.90   
3   PINK VINTAGE PAISLEY PICNIC BAG                       3         2.28   
4            JUMBO BAG PAISLEY PARK                      30         2.88   

          Country  
0  United Kingdom  
1  United Kingdom  
2  United Kingdom  
3  United Kingdom  
4  United Kingdom

### Creating a ratings dataframe by considering the NumberofItemsPurchased as the rating

In [9]:
rdf = df[['TransactionId','ItemCode','NumberOfItemsPurchased']]
rdf

TransactionId  ItemCode  NumberOfItemsPurchased
0             6355745    465549                       6
1             6355745    487494                       3
2             6355745    490686                       3
3             6355745    460593                       3
4             6355745    495201                      30
...               ...       ...                     ...
525873        6235328   1787583                     384
525878        5995847    840336                      72
525879        6260089    434322                       3
525882        6128540    475188                       6
525884        6259946    466137                    3000

[517253 rows x 3 columns]

#### Lets use a sample input-ip to compute the suggestions

In [10]:
ip = [{'ItemCode':1787079,'ItemDescription':'JUMBO BAG RED RETROSPOT','NumberOfItemsPurchased':15},
      {'ItemCode':465549,'ItemDescription':'FAMILY ALBUM WHITE PICTURE FRAME','NumberOfItemsPurchased':6},
      {'ItemCode':475839,'ItemDescription':'LUNCH BOX I LOVE LONDON','NumberOfItemsPurchased':10},
     ]

In [11]:
dfi = pd.DataFrame(ip)

In [12]:
dfi

ItemCode                   ItemDescription  NumberOfItemsPurchased
0   1787079           JUMBO BAG RED RETROSPOT                      15
1    465549  FAMILY ALBUM WHITE PICTURE FRAME                       6
2    475839           LUNCH BOX I LOVE LONDON                      10

#### selecting the transactions which have the i/p items in them

In [13]:
rdf[rdf['ItemCode'].isin(dfi['ItemCode'].values)]

TransactionId  ItemCode  NumberOfItemsPurchased
0             6355745    465549                       6
46            6283376   1787079                       3
936           6162981   1787079                       6
970           6387425   1787079                      30
1048          6199061    475839                       6
...               ...       ...                     ...
525462        6178513    475839                       3
525532        6053388   1787079                     600
525561        6194903    475839                     960
525739        5904085    475839                     384
525810        6288733   1787079                       3

[4923 rows x 3 columns]

In [14]:
#select the transactions which have the i/p items in them
usub = rdf[rdf['ItemCode'].isin(dfi['ItemCode'].values)]
usub

TransactionId  ItemCode  NumberOfItemsPurchased
0             6355745    465549                       6
46            6283376   1787079                       3
936           6162981   1787079                       6
970           6387425   1787079                      30
1048          6199061    475839                       6
...               ...       ...                     ...
525462        6178513    475839                       3
525532        6053388   1787079                     600
525561        6194903    475839                     960
525739        5904085    475839                     384
525810        6288733   1787079                       3

[4923 rows x 3 columns]

In [15]:
usub['TransactionId'].value_counts()

6395829    5
6309435    5
5935545    5
6330247    5
5998663    5
          ..
5932564    1
6375149    1
6314616    1
6171484    1
6322316    1
Name: TransactionId, Length: 3376, dtype: int64

In [16]:
usub.groupby('TransactionId').get_group(6330247)

TransactionId  ItemCode  NumberOfItemsPurchased
39644        6330247   1787079                       6
39809        6330247    465549                       3
39852        6330247    475839                       3
39913        6330247   1787079                      18
40102        6330247   1787079                      12

#### sort the transactions by the length of the usersubset group where transactions with most nos. of i/p items at the top 

In [17]:
 subset = sorted(usub.groupby('TransactionId'),  key=lambda x: len(x[1]), reverse=True)

In [18]:
subset[0:3]

[(5935545,
         TransactionId  ItemCode  NumberOfItemsPurchased
  56875        5935545   1787079                       6
  56898        5935545   1787079                       3
  56931        5935545    475839                       3
  56989        5935545    465549                       3
  57027        5935545   1787079                       3),
 (5998663,
         TransactionId  ItemCode  NumberOfItemsPurchased
  67791        5998663    475839                       3
  67812        5998663   1787079                       9
  67841        5998663    465549                       3
  67887        5998663   1787079                       3
  67904        5998663   1787079                       6),
 (6212987,
        TransactionId  ItemCode  NumberOfItemsPurchased
  2442        6212987    465549                       6
  2475        6212987   1787079                       9
  2476        6212987   1787079                      15
  2514        6212987   1787079                       3

In [19]:
len(subset)

3376

#### lets select the top 100 transactions to base our suggestions on

In [20]:
subset2 = subset[0:100]

## Using Pearson correlation b/w input Transaction and usersubset

In [21]:
from math import sqrt

In [22]:
pearson_dict={}

for Tid, group in subset2:
    group = group.sort_values(by='ItemCode') #lets sort the values first so that there's no confusion later
    inputid = dfi.sort_values(by='ItemCode')
    
    #the n value for the formula
    nrate = len(group)
    
    #lets put the common items b/w the i/p and the user subset in a temporary df
    temp_df = inputid[inputid['ItemCode'].isin(group['ItemCode'].tolist())]
    temp_rate = temp_df['NumberOfItemsPurchased'].values
    temp_grp = group['NumberOfItemsPurchased'].values
    
    #lets calculate the pearson correlation
    Sxx = sum([i**2 for i in temp_rate]) - pow(sum(temp_rate),2)/float(nrate)
    Syy = sum([i**2 for i in temp_grp]) - pow(sum(temp_grp),2)/float(nrate)
    Sxy = sum( i*j for i, j in zip(temp_rate, temp_grp)) - sum(temp_rate)*sum(temp_grp)/float(nrate)
    
    #If the denominator is different than zero, then divide, else, 0 correlation.
    if Sxx != 0 and Syy != 0:
        pearson_dict[Tid] = Sxy/sqrt(Sxx*Syy)
    else:
        pearson_dict[Tid] = 0

In [23]:
pearson_dict.items()

dict_items([(5935545, 0.7572712299036656), (5998663, 0.49050522846032873), (6212987, -0.17019095969632492), (6309435, -0.29829450110074746), (6330247, -0.724260862197483), (6395829, 0), (5902512, 0), (5909607, 0.9169493006161777), (5916053, -0.19619314913492367), (5918440, 0.9622504486493763), (5923764, -0.04550157551932901), (5933807, 0.15600841111499192), (5933961, 0), (5933983, 0.0), (5946061, 0.36650833306891567), (5958656, 0.2364331218717302), (5971966, 0.9169493006161777), (5975002, -0.8518518518518519), (5978137, 0.761840059364464), (5994208, 0), (5996650, -0.07053456158585983), (6008871, 0), (6010466, -0.8551861104941366), (6051584, 0.0), (6074552, -0.6591978953849782), (6081031, -0.345547370232544), (6108025, 0.9169493006161777), (6111941, -0.4937419311010188), (6117188, -0.5555555555555556), (6146547, -0.2455018600084963), (6173299, 0.9169493006161777), (6173387, -0.2364331218717302), (6219356, 0.9169493006161777), (6225087, 0.7777777777777778), (6225109, -0.5555555555555556)

### lets look at the similarityindex for these 100 Transactions

In [24]:
pearsondf = pd.DataFrame.from_dict(pearson_dict, orient='index')
pearsondf.columns = ['similarityIndex']
pearsondf['TransactionId'] = pearsondf.index
pearsondf.index = range(len(pearsondf))

In [25]:
pearsondf

similarityIndex  TransactionId
0          0.757271        5935545
1          0.490505        5998663
2         -0.170191        6212987
3         -0.298295        6309435
4         -0.724261        6330247
..              ...            ...
95         0.500000        6000060
96         0.000000        6000071
97         0.000000        6000830
98         0.000000        6001380
99        -0.785714        6002557

[100 rows x 2 columns]

#### top 50 Transactions with highest similarity index

In [26]:
toptran=pearsondf.sort_values(by='similarityIndex', ascending=False)[0:50]
toptran.head()

similarityIndex  TransactionId
53         1.000000        5915723
85         1.000000        5982141
9          0.962250        5918440
30         0.916949        6173299
26         0.916949        6108025

#### merge the top 50 transactions with the rating df(rdf) to get the item codes and their rating(Numberofitemspurchased)

In [27]:
toptranrate = toptran.merge(rdf, left_on='TransactionId', right_on='TransactionId', how='inner')
toptranrate.rename(columns = {'NumberOfItemsPurchased':'Rating'},inplace=True)
toptranrate.head()

similarityIndex  TransactionId  ItemCode  Rating
0              1.0        5915723   1787079     600
1              1.0        5915723    460656       9
2              1.0        5915723    470106     600
3              1.0        5915723    475902       9
4              1.0        5915723    435267       9

In [28]:
df[df['TransactionId'] == 5915723].head()

UserId  TransactionId               TransactionTime  ItemCode  \
377962  315672        5915723  Mon Feb 19 08:57:00 IST 2018   1787079   
377963  315672        5915723  Mon Feb 19 08:57:00 IST 2018    460656   
377964  315672        5915723  Mon Feb 19 08:57:00 IST 2018    470106   
377965  315672        5915723  Mon Feb 19 08:57:00 IST 2018    475902   
377966  315672        5915723  Mon Feb 19 08:57:00 IST 2018    435267   

                       ItemDescription  NumberOfItemsPurchased  CostPerItem  \
377962  JUMBO  BAG BAROQUE BLACK WHITE                     600         2.28   
377963        RED RETROSPOT PICNIC BAG                       9         4.08   
377964         JUMBO BAG PINK POLKADOT                     600         2.28   
377965     LUNCH BAG DOLLY GIRL DESIGN                       9         2.28   
377966         LUNCH BAG  BLACK SKULL.                       9         2.28   

               Country  
377962  United Kingdom  
377963  United Kingdom  
377964  United Kingdom  
377965  United Kingdom  
377966  United Kingdom

In [29]:
toptranrate.shape

(9863, 4)

#### lets calculate the weighted rating

In [30]:
toptranrate['weightedrating'] = toptranrate['similarityIndex']* toptranrate['Rating']
toptranrate.head()

similarityIndex  TransactionId  ItemCode  Rating  weightedrating
0              1.0        5915723   1787079     600           600.0
1              1.0        5915723    460656       9             9.0
2              1.0        5915723    470106     600           600.0
3              1.0        5915723    475902       9             9.0
4              1.0        5915723    435267       9             9.0

#### Applying a sum to the toptransactions after grouping it up by itemcode

In [31]:
temptopTransactionrating = toptranrate.groupby('ItemCode').sum()[['similarityIndex','weightedrating']]
temptopTransactionrating.columns = ['sum_similarityIndex','sum_weightedRating']
temptopTransactionrating.head()

sum_similarityIndex  sum_weightedRating
ItemCode                                         
63                   0.000000            0.000000
1596                 0.654654            3.927922
2520                 0.000000            0.000000
2940                 0.896258            2.688774
210042               2.404242            9.484541

In [32]:
temptopTransactionrating.shape

(2237, 2)

In [33]:
#Creates an empty dataframe
recommendation_df = pd.DataFrame()
#Now we take the weighted average
recommendation_df['weighted average recommendation score'] = temptopTransactionrating['sum_weightedRating']/temptopTransactionrating['sum_similarityIndex']
recommendation_df['ItemCode'] = temptopTransactionrating.index
recommendation_df.head()

weighted average recommendation score  ItemCode
ItemCode                                                 
63                                          NaN        63
1596                                   6.000000      1596
2520                                        NaN      2520
2940                                   3.000000      2940
210042                                 3.944919    210042

### The top 10 item codes recommended with the highest weighted average recommendation score

In [34]:
recommendation_df = recommendation_df.sort_values(by='weighted average recommendation score', ascending=False)
recommendation_df.head(10)

weighted average recommendation score  ItemCode
ItemCode                                                 
446145                               288.000000    446145
490203                               237.494376    490203
434931                               217.532663    434931
434154                               216.000000    434154
460488                               170.841896    460488
470106                               169.595259    470106
470085                               121.563054    470085
1787079                               98.957698   1787079
434973                                94.670667    434973
1775928                               79.123859   1775928

In [35]:
recommendation_df.head(10)['ItemCode'].values

array([ 446145,  490203,  434931,  434154,  460488,  470106,  470085,
       1787079,  434973, 1775928], dtype=int64)

In [36]:
frequently_bought=df[df['ItemCode'].isin(recommendation_df.head(10)['ItemCode'].values)]
frequently_bought

UserId  TransactionId               TransactionTime  ItemCode  \
22      278166        6355745  Sat Feb 02 12:50:00 IST 2019    490203   
46      337701        6283376  Wed Dec 26 09:06:00 IST 2018   1787079   
237         -1        6143225  Mon Sep 10 11:58:00 IST 2018    434931   
244         -1        6143225  Mon Sep 10 11:58:00 IST 2018    434973   
333         -1        6143225  Mon Sep 10 11:58:00 IST 2018    490203   
...        ...            ...                           ...       ...   
525385  300258        6169273  Tue Oct 02 08:27:00 IST 2018    434973   
525522  331149        6101810  Tue Aug 07 06:37:00 IST 2018    470106   
525532  357609        6053388  Fri Jun 29 06:57:00 IST 2018   1787079   
525555  321636        5954036  Sat Mar 31 07:34:00 IST 2018    434973   
525810      -1        6288733  Sun Dec 30 11:07:00 IST 2018   1787079   

                       ItemDescription  NumberOfItemsPurchased  CostPerItem  \
22        JUMBO BAG VINTAGE CHRISTMAS                       18         2.88   
46             JUMBO BAG RED RETROSPOT                       3         2.88   
237                    JUMBO BAG TOYS                        3         5.70   
244                     JUMBO BAG OWLS                      21         5.70   
333       JUMBO BAG VINTAGE CHRISTMAS                       12         5.70   
...                                ...                     ...          ...   
525385                  JUMBO BAG OWLS                     180         2.28   
525522         JUMBO BAG PINK POLKADOT                     900         2.48   
525532         JUMBO BAG RED RETROSPOT                     600         2.48   
525555                  JUMBO BAG OWLS                     300         2.28   
525810  JUMBO  BAG BAROQUE BLACK WHITE                       3         2.88   

               Country  
22      United Kingdom  
46      United Kingdom  
237     United Kingdom  
244     United Kingdom  
333     United Kingdom  
...                ...  
525385  United Kingdom  
525522  United Kingdom  
525532  United Kingdom  
525555  United Kingdom  
525810  United Kingdom  

[8689 rows x 8 columns]

In [37]:
frequently_bought['ItemCode'].nunique()

10

In [38]:
frequently_bought.drop_duplicates('ItemCode')

UserId  TransactionId               TransactionTime  ItemCode  \
22     278166        6355745  Sat Feb 02 12:50:00 IST 2019    490203   
46     337701        6283376  Wed Dec 26 09:06:00 IST 2018   1787079   
237        -1        6143225  Mon Sep 10 11:58:00 IST 2018    434931   
244        -1        6143225  Mon Sep 10 11:58:00 IST 2018    434973   
346        -1        6143225  Mon Sep 10 11:58:00 IST 2018    470085   
918    345954        6162981  Fri Sep 28 10:51:00 IST 2018    470106   
959    339822        6255403  Mon Dec 10 09:23:00 IST 2018    460488   
1407       -1        6288843  Sun Dec 30 12:58:00 IST 2018    434154   
1834       -1        6388019  Sat Feb 16 13:24:00 IST 2019   1775928   
11921      -1        5955664  Sat Mar 31 13:57:00 IST 2018    446145   

                           ItemDescription  NumberOfItemsPurchased  \
22            JUMBO BAG VINTAGE CHRISTMAS                       18   
46                 JUMBO BAG RED RETROSPOT                       3   
237                        JUMBO BAG TOYS                        3   
244                         JUMBO BAG OWLS                      21   
346              JUMBO BAG SPACEBOY DESIGN                       3   
918                JUMBO BAG PINK POLKADOT                       3   
959    JUMBO BAG SCANDINAVIAN BLUE PAISLEY                     600   
1407                   GREEN POLKADOT BOWL                       6   
1834       GIRLS ALPHABET IRON ON PATCHES                       48   
11921                GREEN POLKADOT PLATE                        6   

       CostPerItem         Country  
22            2.88  United Kingdom  
46            2.88  United Kingdom  
237           5.70  United Kingdom  
244           5.70  United Kingdom  
346           5.70  United Kingdom  
918           2.88  United Kingdom  
959           2.48  United Kingdom  
1407          3.40  United Kingdom  
1834          0.58  United Kingdom  
11921         4.55  United Kingdom

In [39]:
recomm = frequently_bought[frequently_bought['ItemCode'] == 1787079 ]
recomm

UserId  TransactionId               TransactionTime  ItemCode  \
46      337701        6283376  Wed Dec 26 09:06:00 IST 2018   1787079   
936     345954        6162981  Fri Sep 28 10:51:00 IST 2018   1787079   
970     328440        6387425  Sat Feb 16 10:35:00 IST 2019   1787079   
1298        -1        6288843  Sun Dec 30 12:58:00 IST 2018   1787079   
1511        -1        6288843  Sun Dec 30 12:58:00 IST 2018   1787079   
...        ...            ...                           ...       ...   
525245  381654        6384081  Wed Feb 13 12:56:00 IST 2019   1787079   
525327  274428        6374159  Sun Feb 10 11:00:00 IST 2019   1787079   
525329  378357        6137340  Tue Sep 04 11:48:00 IST 2018   1787079   
525532  357609        6053388  Fri Jun 29 06:57:00 IST 2018   1787079   
525810      -1        6288733  Sun Dec 30 11:07:00 IST 2018   1787079   

                       ItemDescription  NumberOfItemsPurchased  CostPerItem  \
46             JUMBO BAG RED RETROSPOT                       3         2.88   
936            JUMBO BAG RED RETROSPOT                       6         2.88   
970            JUMBO BAG RED RETROSPOT                      30         2.88   
1298              JUMBO BAG STRAWBERRY                       3         5.70   
1511           JUMBO BAG RED RETROSPOT                      15         5.70   
...                                ...                     ...          ...   
525245         JUMBO BAG RED RETROSPOT                     150         2.88   
525327  JUMBO  BAG BAROQUE BLACK WHITE                     300         2.48   
525329         JUMBO BAG RED RETROSPOT                     300         2.48   
525532         JUMBO BAG RED RETROSPOT                     600         2.48   
525810  JUMBO  BAG BAROQUE BLACK WHITE                       3         2.88   

               Country  
46      United Kingdom  
936     United Kingdom  
970     United Kingdom  
1298    United Kingdom  
1511    United Kingdom  
...                ...  
525245  United Kingdom  
525327  United Kingdom  
525329  United Kingdom  
525532  United Kingdom  
525810  United Kingdom  

[3844 rows x 8 columns]

In [40]:
recomm[recomm['ItemDescription'] !=  'JUMBO BAG RED RETROSPOT']

UserId  TransactionId               TransactionTime  ItemCode  \
1298        -1        6288843  Sun Dec 30 12:58:00 IST 2018   1787079   
2059        -1        6388019  Sat Feb 16 13:24:00 IST 2019   1787079   
2476        -1        6212987  Sun Nov 11 08:02:00 IST 2018   1787079   
2514        -1        6212987  Sun Nov 11 08:02:00 IST 2018   1787079   
3059    281568        6344448  Mon Jan 28 09:13:00 IST 2019   1787079   
...        ...            ...                           ...       ...   
523894  336609        5952155  Tue Mar 27 10:34:00 IST 2018   1787079   
524400  304017        5984869  Sun Apr 29 11:44:00 IST 2018   1787079   
524623  316638        6093274  Wed Aug 01 06:31:00 IST 2018   1787079   
525327  274428        6374159  Sun Feb 10 11:00:00 IST 2019   1787079   
525810      -1        6288733  Sun Dec 30 11:07:00 IST 2018   1787079   

                       ItemDescription  NumberOfItemsPurchased  CostPerItem  \
1298              JUMBO BAG STRAWBERRY                       3         5.70   
2059    JUMBO  BAG BAROQUE BLACK WHITE                       6         5.70   
2476              JUMBO BAG STRAWBERRY                      15         5.70   
2514    JUMBO  BAG BAROQUE BLACK WHITE                       3         5.70   
3059    JUMBO  BAG BAROQUE BLACK WHITE                     300         2.48   
...                                ...                     ...          ...   
523894            JUMBO BAG STRAWBERRY                     600         2.28   
524400  JUMBO  BAG BAROQUE BLACK WHITE                     300         2.28   
524623            JUMBO BAG STRAWBERRY                      30         2.88   
525327  JUMBO  BAG BAROQUE BLACK WHITE                     300         2.48   
525810  JUMBO  BAG BAROQUE BLACK WHITE                       3         2.88   

               Country  
1298    United Kingdom  
2059    United Kingdom  
2476    United Kingdom  
2514    United Kingdom  
3059    United Kingdom  
...                ...  
523894  United Kingdom  
524400  United Kingdom  
524623  United Kingdom  
525327  United Kingdom  
525810  United Kingdom  

[1755 rows x 8 columns]

## Final function to generate frequently bought products

In [41]:
#final function to generate frequently bought products
def predict(df,ip):
    rdf = df[['TransactionId','ItemCode','NumberOfItemsPurchased']]
    dfi = pd.DataFrame(ip)
    #select the transactions which have the i/p items in them
    usub = rdf[rdf['ItemCode'].isin(dfi['ItemCode'].values)]
    #sort the transactions by the length of the usersubset group where transactions with most nos. of i/p items at the top 
    subset = sorted(usub.groupby('TransactionId'),  key=lambda x: len(x[1]), reverse=True)
    
    
    pearson_dict={}

    for Tid, group in subset2:
        group = group.sort_values(by='ItemCode') #lets sort the values first so that there's no confusion later
        inputid = dfi.sort_values(by='ItemCode')
    
        #the n value for the formula
        nrate = len(group)
    
    #lets put the common items b/w the i/p and the user subset in a temporary df
        temp_df = inputid[inputid['ItemCode'].isin(group['ItemCode'].tolist())]
        temp_rate = temp_df['NumberOfItemsPurchased'].values
        temp_grp = group['NumberOfItemsPurchased'].values
    
    #lets calculate the pearson correlation
        Sxx = sum([i**2 for i in temp_rate]) - pow(sum(temp_rate),2)/float(nrate)
        Syy = sum([i**2 for i in temp_grp]) - pow(sum(temp_grp),2)/float(nrate)
        Sxy = sum( i*j for i, j in zip(temp_rate, temp_grp)) - sum(temp_rate)*sum(temp_grp)/float(nrate)
    
    #If the denominator is different than zero, then divide, else, 0 correlation.
        if Sxx != 0 and Syy != 0:
            pearson_dict[Tid] = Sxy/sqrt(Sxx*Syy)
        else:
            pearson_dict[Tid] = 0
    
    pearsondf = pd.DataFrame.from_dict(pearson_dict, orient='index')
    pearsondf.columns = ['similarityIndex']
    pearsondf['TransactionId'] = pearsondf.index
    pearsondf.index = range(len(pearsondf))
    toptran=pearsondf.sort_values(by='similarityIndex', ascending=False)[0:50]
    toptranrate = toptran.merge(rdf, left_on='TransactionId', right_on='TransactionId', how='inner')
    toptranrate.rename(columns = {'NumberOfItemsPurchased':'Rating'},inplace=True)
    
    toptranrate['weightedrating'] = toptranrate['similarityIndex']* toptranrate['Rating']
    #Applies a sum to the toptransactions after grouping it up by itemcode
    temptopTransactionrating = toptranrate.groupby('ItemCode').sum()[['similarityIndex','weightedrating']]
    temptopTransactionrating.columns = ['sum_similarityIndex','sum_weightedRating']
    
    #Creates an empty dataframe
    recommendation_df = pd.DataFrame()
    #Now we take the weighted average
    recommendation_df['weighted average recommendation score'] = temptopTransactionrating['sum_weightedRating']/temptopTransactionrating['sum_similarityIndex']
    recommendation_df['ItemCode'] = temptopTransactionrating.index
    recommendation_df = recommendation_df.sort_values(by='weighted average recommendation score', ascending=False)
    result = recommendation_df.head(2)['ItemCode'].tolist()
    return result

In [46]:
suggestions = predict(df,[{'ItemCode':1787079,'ItemDescription':'JUMBO BAG RED RETROSPOT','NumberOfItemsPurchased':15}])
print('Frequently bought together itemcodes:',suggestions)

Frequently bought together itemcodes: [445578, 478884]


In [49]:
df.groupby('ItemCode').get_group(445578)

UserId  TransactionId               TransactionTime  ItemCode  \
5246        -1        6390813  Sun Feb 17 12:26:00 IST 2019    445578   
7683        -1        6104560  Wed Aug 08 11:18:00 IST 2018    445578   
8776    275877        6373389  Sun Feb 10 08:09:00 IST 2019    445578   
10402       -1        5921839  Wed Feb 21 10:59:00 IST 2018    445578   
11020       -1        6052332  Wed Jun 27 06:41:00 IST 2018    445578   
...        ...            ...                           ...       ...   
517136  360633        6303902  Tue Jan 08 10:21:00 IST 2019    445578   
517683  360633        6180889  Fri Oct 12 11:26:00 IST 2018    445578   
517826  352359        5949427  Mon Mar 26 06:20:00 IST 2018    445578   
518267  381507        6210886  Wed Nov 07 07:40:00 IST 2018    445578   
522024  319935        5974463  Tue Apr 17 11:27:00 IST 2018    445578   

               ItemDescription  NumberOfItemsPurchased  CostPerItem  \
5246    RED SPOTTY BISCUIT TIN                       9        11.45   
7683    RED SPOTTY BISCUIT TIN                       3        11.45   
8776    RED SPOTTY BISCUIT TIN                       3         5.18   
10402   RED SPOTTY BISCUIT TIN                       3        11.69   
11020   RED SPOTTY BISCUIT TIN                       3        11.45   
...                        ...                     ...          ...   
517136  RED SPOTTY BISCUIT TIN                      36         4.49   
517683  RED SPOTTY BISCUIT TIN                      18         4.49   
517826  RED SPOTTY BISCUIT TIN                      72         4.49   
518267  RED SPOTTY BISCUIT TIN                      18         4.49   
522024  RED SPOTTY BISCUIT TIN                      18         5.18   

               Country  
5246    United Kingdom  
7683    United Kingdom  
8776    United Kingdom  
10402   United Kingdom  
11020   United Kingdom  
...                ...  
517136  United Kingdom  
517683  United Kingdom  
517826  United Kingdom  
518267  United Kingdom  
522024  United Kingdom  

[394 rows x 8 columns]

#### to check whether the users who have bought the input item have also bought the suggested items 

In [44]:
id1 = df.groupby('ItemCode').get_group(445578)['TransactionId']
id2 = df.groupby('ItemCode').get_group(1787079)['TransactionId']  

In [45]:
len(id1[id1.isin(id2.to_list())])

194

#### There are 194 common Transactions made for this particular item which suggests that users who bought one item also bought the other item

#### The main Idea to note here is that the suggestions are based on the weigted average rating and not on the frequency of the transactionid's, where common transactions for an item maybe few but if it has been rated very highly then that item would be suggested first